pkg/volume/csi/csi_attacher.go

```diff
@@ -139,31 +153,118 @@ func (c *csiAttacher) WaitForAttach(spec *volume.Spec, _ string, pod *v1.Pod, ti
-	attach, err := c.k8s.StorageV1().VolumeAttachments().Get(context.TODO(), attachID, meta.GetOptions{})
+
+	// Check the lister to see if the object exists before watching for it, otherwise you may encounter watch timeouts
+	// because any relevant events may have happened between subsequent runs of Attach().
+	attach, err := c.plugin.volumeAttachmentLister.Get(attachID)
 	if err != nil {
-		klog.Error(log("attacher.WaitForAttach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
+		klog.Error(log("attacher.WaitForAttach failed for volume [%v] (will continue to try): %v", volumeHandle, err))
 		return "", fmt.Errorf("volume %v has GET error for volume attachment %v: %v", volumeHandle, attachID, err)
 	}
-	err = c.waitForVolumeAttachDetachStatus(attach, volumeHandle, attachID, timer, timeout, verifyAttachmentStatus)
+
+	err = c.waitForVolumeAttachDetachStatusWithInformer(attach, volumeHandle, attachID, timeout, verifyAttachmentStatus)
 	if err != nil {
 		return "", err
 	}
-	return attach.Name, nil
+	return attachID, nil
+}
+
+func (c *csiAttacher) waitForVolumeDetachmentWithInformer(volumeHandle, attachID string, timeout time.Duration) error {
+	klog.V(4).Info(log("probing VolumeAttachment [id=%v]", attachID))
+
+	// Check the lister to see if the object exists before watching for it, otherwise you may encounter watch timeouts
+	// because any relevant events may have happened between subsequent runs of Detach().
+	attach, err := c.plugin.volumeAttachmentLister.Get(attachID)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] not found, object deleted", attachID, volumeHandle))
+			return nil
+		}
+		return errors.New(log("detacher.WaitForDetach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
+	}
+
+	return c.waitForVolumeAttachDetachStatusWithInformer(attach, volumeHandle, attachID, timeout, verifyDetachmentStatus)
+}
+
+func (c *csiAttacher) waitForVolumeAttachDetachStatusWithInformer(attach *storage.VolumeAttachment, volumeHandle, attachID string, timeout time.Duration, verifyStatus verifyAttachDetachStatus) error {
+	successful, err := verifyStatus(attach, volumeHandle)
+	if err != nil {
+		return err
+	}
+	if successful {
+		return nil
+	}
+
+	ch := c.plugin.volumeAttachmentInformerHook.GetWatch(attachID)
+	if ch == nil {
+		klog.Error(log("[attachment.ID=%s] watch channel could not be located", attachID))
+		return errors.New("volume attachment watch channel could not be located")
+	}
+
+	timer := time.NewTimer(timeout) // TODO (vladimirvivien) investigate making this configurable
+	defer timer.Stop()
+
+	for {
+		select {
+		// Ch is a buffered channel that receives events from the SharedInformerHook. Since there is the
+		// possibility of events being dropped, we ignore the received event and treat it as a notification
+		// to check the lister for the latest status.
+		case _, ok := <-ch:
```

Comment:
so you are still using event based monitor? The order option is periodically get from lister so that you need to create watch?

Reply:
I wanted to try the event-based monitor approach first because there were concerns with performance regressions if we went with a polling model (assuming that meant we would block for some time and repeatedly poll the lister until it was available).

With that being said, the way attach currently works requires the VolumeAttachment object to exist in the lister prior to watching it for an "updated" event. Given this design, what most often happens is:
- `Attach()` runs and creates a VolumeAttachment
- `waitForVolumeAttachmentInternal()` runs, but does not find the VolumeAttachment in the lister (because it was just created, not enough time has passed yet), it returns an error (so no watch gets created)
- `Attach()` runs again and creates a VolumeAttachment and fails (because it exists)
- `waitForVolumeAttachmentInternal()` runs, this time it finds the VolumeAttachment in the lister
- `waitForVolumeAttachDetachStatus()` runs, it sees the VolumeAttachment succeeded and returns (without ever needing to create a watch)

With the way `Attach()` is repeatedly called by the AttachDetachController, we do typically end up repeatedly polling for the VolumeAttachment. If we got rid of the watch logic entirely and just let the ADController repeatedly try until it succeeds we could have more async behavior.

Was what I described above in line with your thoughts on a polling solution?

Reply:
Thanks a lot for the detailed explanation. For step 2 and 3, how long it typically take? Another word, how long it takes for the VA object get into the lister?

Reply:
Running one of the e2e tests against a kind cluster, I observed it took 55552 nanoseconds ~= 55 microseconds (in the case that a VolumeAttachment was immediately fetched from the lister after creation).

Occasionally the object is not immediately available, in which case we have to wait 100ms for the reconciler to run again.

Reply:
got it, thanks!

I think it is find to let it retry here.

Reply:
I am trying to think it again about periodic pull approach. Since it is pulling from Lister, what performance regressions you are concern? In think overhead should be small. The time period to pull is set to every say a few second seems fine to me.  
The watch approach is nice too, just in case some race condition, the watch could not get the event because it was deleted just before you start to watch it, then you have to wait until watch timeout which might cause a much longer time.

Reply:
I ran the volume stress test a couple times, and observed ~81% of the time we verify attachment during the watch and not via the get from the lister, however the test fails saying not all volumes are bound in time. Comparing this against the base revision now.

I thought there may be performance regressions in CPU because the reconciler would have to call Attach more than once for an attachment to complete. I also thought because this might impact how long it takes for volumes to be attached, however since the reconciler runs 10 times per second I wonder if this impact would even be noticeable.

Reply:
@jingxu97 @msau42 

Below is an example of a more asynchronous version of `Attach()` that is designed to be called repeatedly by the reconciler until it succeeds. If we go this route it should cut down on complexity and lines of code. Thoughts on this approach instead?

https://github.com/chrishenzie/kubernetes/blob/volume-attachment-informer-retry/pkg/volume/csi/csi_attacher.go#L66-L142

Reply:
It certainly makes simplifies things, with the main downside that we will retry Attach() multiple times. Do you have a spreadsheet you can share that has the perf comparison?

@jsafrane do you remember why we originally implemented a Watch for the Attach operation? PV controller and resize controller are async operations for CSI as well. It does mess up our metrics a little bit, but they're already broken for provision and resize.

Reply:
Watch was IMO used only to speed things up and simulate existing behavior of in-tree attach. Volume attachment is part of pod startup and it might be watched by scale team. Will the reconciler add exponential backoff between each Attach call? It grows quickly and it could affect attach speed.

Reply:
Yes the reconciler does exponentially back off. When I ran the stress test I saw it wait 500ms between tries.

Reply:
Was this with the mock driver? Could you try injecting a sleep of various magnitudes to the mock driver's attach call? (maybe 1s, 10, 30). I believe with the operation_executor, there is an additional backoff that could get bigger on the order of a minute or more.